### PR TITLE
Feature/api query create

### PR DESCRIPTION
**Pull request by ulferts**
As of now, this PR is a stub to enable moving the front end to rely on api v3 only. The end points are by no means finished.

Because the endpoints did not yet accept parameters (`project` in particular), there is a project scoped end point as well as a global endpoint. Once parameters are accepted, the project scoped endpoint can be removed.

https://community.openproject.com/projects/openproject/work_packages/24491

### CODE REVIEW

**Comment by ulferts** [comment]
#### Empty POST at /api/v3/queries/form 
When issuing a POST against the /api/v3/queries/form endpoint without a body, this is what is returned within the payload part of the response (omitting schema and errors for brevity):

```
{
      "_type": "Query",
      "name": "",
      "filters": [],
      "public": false,
      "sums": false,
      "starred": false,
      "_links": {
        "self": {
          "href": "/api/v3/queries/default",
          "title": ""
        },
        "results": {
          "href": "/api/v3/work_packages?offset=1&pageSize=2"
        },
        "columns": [
          {
            "href": "/api/v3/queries/columns/id",
            "title": "ID"
          },
          {
            "href": "/api/v3/queries/columns/project",
            "title": "Project"
          },
          {
            "href": "/api/v3/queries/columns/subject",
            "title": "Subject"
          },
          {
            "href": "/api/v3/queries/columns/type",
            "title": "Type"
          },
          {
            "href": "/api/v3/queries/columns/status",
            "title": "Status"
          },
          {
            "href": "/api/v3/queries/columns/priority",
            "title": "Priority"
          },
          {
            "href": "/api/v3/queries/columns/assignee",
            "title": "Assignee"
          },
          {
            "href": "/api/v3/queries/columns/updatedAt",
            "title": "Updated on"
          }
        ],
        "groupBy": {
          "href": null,
          "title": null
        },
        "sortBy": [],
        "schema": {
          "href": "/api/v3/queries/schema"
        },
        "update": {
          "href": "/api/v3/queries/form",
          "method": "post"
        },
        "user": {
          "href": null
        },
        "project": {
          "href": null
        }
      }
```

The following parts are wrong here IMO:
* The following fields should not be present as they are not writable. The idea of the payload object is to provide an object to the client which he can amend according to the schema and finally send to the create endpoint. That object must thus be valid if no errors are noted. 
  * '_type'
  * 'user' link 
  * 'self' link
  * 'results' link
  * 'schema' link
  * 'update' link
  * 'user' link (should by the way be set to the current user)
* The following fields do not have default values set (It might make sense to use `Query.new_default` here):
  * 'sort by' should be parent (desc) by default
  * 'filters' which should filter by status - open.
  * Interestingly, both fields are set to the default values when calling the form endpoint nested inside a project. 
 
#### POST at /api/v3/queries/form with content
* It is not possible to alter the values of the following fields (a method not found error is thrown):
  * 'public'
  * 'starred' 
* It is possible to provide faulty values without an error being reported:
  * 'sortBy'
  * 'columns'
  * 'groupBy'
* It is not possible to alter the project. Of course one can alternatively rely on the /api/v3/projects/:project_id/queries endpoint but that way we would need to keep that endpoint which we would not have to otherwise. 

#### POST at /api/v3/queries with content
* same errors on 'public' and 'starred' (same as form)
* No error messages on faulty values (same as form)
* It is possible to select an arbitrary user to be the querie's user. The use should always be the current user.
* project can not be set here as well
* I didn't double check, but there does not seem to be a permission check regarding the `public` property which should require the `manage public queries` permission if the client wants to set it to true.
* I didn't see other permission checks but they might be there.

**Comment by ulferts** [comment]
Haven't yet taken a look at the code but focused on testing instead. My previous remarks seem to have been solved mostly.

Only two things I noticed:
* Setting the project is only possible via the id (e.g. `/api/v3/projects/1`) and not via the identifier (e.g. `/api/v3/projects/seeded_project`). This is problematic as the client might not have the id based href as the backend mostly issues out identifier based hrefs. 
* It is possible to provide bogus values via the group-by href (e.g. `/api/v3/queries/group_bys/project-jksdjfldskjffs`) which causes the backend to fall back to ascending order. This is not a major problem IMO.

Will take a look at the code later on.

File: app/contracts/queries/create_contract.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#++
+
+require 'queries/base_contract'
+
+module Queries
+  class CreateContract < BaseContract
+    validate :user_allowed_to_make_public
+
+    def user_allowed_to_make_public
+      if is_public && !user.allowed_to?(:manage_public_queries, model.project)
```
Comment (ulferts) [inline app/contracts/queries/create_contract.rb]: I guess (and it is just a guess at this point) that we will move this part to the `BaseContract` once the update action is being worked on. But we can also do it then.

File: app/models/queries/not_existing_filter.rb [hunk 2]
```diff
@@ -52,4 +52,13 @@ def always_false
...
   def validate_inclusion_of_operator; end
+
+  def to_hash
+    {
+      non_existent_filter: {
+        operator: operator,
+        values: values
+      }
+    }
+  end
```
Comment (ulferts) [inline app/models/queries/not_existing_filter.rb:63]: 👍  Thanks

File: app/models/query.rb [hunk 3]
```diff
@@ -208,6 +209,32 @@ def validate_work_package_filters
...
+        errors.add :sort_criteria, I18n.t(:error_invalid_sort_criterion, value: name)
+      end
+    end
+  end
+
+  def validate_group_by
+    unless group_by.nil? || groupable_columns.map(&:name).map(&:to_s).include?(group_by.to_s)
+      errors.add :group_by, I18n.t(:error_invalid_group_by, value: group_by)
+    end
+  end
```
Comment (ulferts) [inline app/models/query.rb]: Thanks for the groundwork

File: app/contracts/queries/base_contract.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    validate :validate_project
+
+    def initialize(query, user)
+      super query
+
+      @user = user
+    end
+
+    def validate_project
+      errors.add :project, :error_not_found if project_id && !Project.exists?(project_id)
```
Comment (ulferts) [inline app/contracts/queries/base_contract.rb]: Do we check for project visibility (with regards to the current user)? As the POST endpoint is not project nested, we need to check this somewhere below the api endpoint.

File: lib/api/v3/queries/create_form_representer.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          end
+        end
+
+        private
+
+        def allow_commit?
+          represented.name.present? && (
+              (!represented.is_public && current_user.allowed_to?(:save_queries, represented.project)) ||
+              (represented.is_public && current_user.allowed_to?(:manage_public_queries, represented.project))
+            ) && @errors.empty?
```
Comment (ulferts) [inline lib/api/v3/queries/create_form_representer.rb]: This is duplicating the validations. Can't we rely on the validations in the contract and in the model (for as long as we have that duplicity) instead?

File: lib/api/v3/queries/create_query.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          ::API::V3::Queries::QueryRepresenter
+        end
+      end
+
+      def create_query_form(
+        query,
+        current_user:,
+        contract_class: ::Queries::CreateContract,
+        form_class: ::API::V3::Queries::CreateFormRepresenter,
+        action: :update
```
Comment (ulferts) [inline lib/api/v3/queries/create_query.rb]: The `action` is or at least sounds wrong.

File: lib/api/v3/queries/query_payload_representer.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          elsif value.is_a? Array
+            value.map { |v| trim_links v }
+          else
+            value
+          end
+        end
+
+        def convert_attribute(attribute)
+          ::API::Utilities::PropertyNameConverter.from_ar_name(attribute)
+        end
```
Comment (ulferts) [inline lib/api/v3/queries/query_payload_representer.rb]: Is it possible to create a superclass for QueryRepresenter QueryPayloadRepresenter?

File: app/models/query.rb [hunk 8]
```diff
@@ -208,6 +209,32 @@ def validate_work_package_filters
...
+
+    sort_criteria.each do |name, _dir|
+      unless available_criteria.include? name.to_s
+        errors.add :sort_criteria, I18n.t(:error_invalid_sort_criterion, value: name)
+      end
+    end
+  end
+
+  def validate_group_by
+    unless group_by.nil? || groupable_columns.map(&:name).map(&:to_s).include?(group_by.to_s)
```
Comment (ulferts) [inline app/models/query.rb]: group_by seems to be `""` when not grouping and coming from the DB. Thus we would need to check for `blank?` instead.

File: app/contracts/queries/base_contract.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    end
+
+    def validate_project
+      errors.add :project, :error_not_found if project_id.present? && !project_visible?
+    end
+
+    def project_visible?
+      project = Project.find_by_id project_id
+
+      project && project.visible?(user)
```
Comment (ulferts) [inline app/contracts/queries/base_contract.rb]:
```
Project.visible(user).where(id: project_id).exists?
```

Would cover all boolean expressions here and avoids instantiating an object on top.

Comment (ulferts) [reply app/contracts/queries/base_contract.rb]: Come to think of it, checking for visibility is not enough here. A check for `:view_work_packages` would be better I think.

```
Project.allowed_to(user, :view_work_packages).where(id: project_id).exists?
```

Comment (machisuji) [reply app/contracts/queries/base_contract.rb]: Cool thanks for the pointer. Will update it.

File: lib/api/v3/queries/queries_api.rb [hunk 10]
```diff
@@ -88,10 +90,16 @@ def allowed_to?(action)
...
 
+          post do
+            create_query request_body, current_user
+          end
+
           params do
             requires :id, desc: 'Query id'
           end
           route_param :id do
+            mount API::V3::Queries::UpdateFormAPI
```
Comment (ulferts) [inline lib/api/v3/queries/queries_api.rb]: this has to be placed after the `before` block in order for the query to be found

Comment (machisuji) [reply lib/api/v3/queries/queries_api.rb]: Ah ... thanks!

File: app/contracts/queries/base_contract.rb [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    def project_visible?
+      Project.visible(user).where(id: project_id).exists?
+    end
+
+    def user_allowed_to_make_public
+      return if model.project_id.present? && model.project.nil?
+
+      if is_public && !user.allowed_to?(:manage_public_queries, model.project)
```
Comment (ulferts) [inline app/contracts/queries/base_contract.rb]: You need to add `global: model.project.nil?` here to check for non project scoped queries.

File: app/services/queries/update_query_service.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    result, errors = validate_and_save query
+
+    ServiceResult.new success: result, errors: errors, result: query
+  end
+
+  def initialize_contract!(query)
+    self.contract = self.class.contract.new query, user
+  end
+end
```
Comment (ulferts) [inline app/services/queries/update_query_service.rb]: Other than the `update` method, the `UpdateQueryService` and the `CreateQueryService` have equal code. Please extract the common code.

File: docs/api/apiv3/endpoints/queries.apib [hunk 13]
```diff
@@ -1231,6 +1231,53 @@ Returns a collection of queries. The collection can be filtered via query parame
...
+The returned form will be pre-filled with default values for every property, if available.
+
+For more details and all possible responses see the general specification of [Forms](#forms).
+
+## Query Create Form [POST]
+
++ Response 200 (application/hal+json)
+
+    [Example Form][]
+
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/queries.apib:1369]: Only create is described here, update is missing as of now.

File: docs/api/apiv3/endpoints/queries.apib [hunk 14]
```diff
@@ -1231,6 +1231,53 @@ Returns a collection of queries. The collection can be filtered via query parame
...
+    Returned when the client tries to use (link) unknown users, projects or operators in the query to be created.
+
+    + Body
+
+            {
+                "_type": "Error",
+                "errorIdentifier": "urn:openproject-org:api:v3:errors:NotFound",
+                "message": "User 42 not found"
+            }
+
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/queries.apib]: A `422` error will additionally be returned when invalid references are provided (e.g. group by) or other validations (e.g. name) are not met.

File: lib/api/v3/queries/update_form_api.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+              UpdateFormRepresenter.new query, current_user: current_user, errors: api_errors
+            else
+              fail ::API::Errors::MultipleErrors.create_if_many(api_errors)
+            end
+          end
+        end
+      end
+    end
+  end
+end
```
Comment (ulferts) [inline lib/api/v3/queries/update_form_api.rb:46]: Almost the same as CreateFormAPI. Any chance to share the code?

File: lib/api/v3/queries/update_form_representer.rb [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        private
+
+        def allow_commit?
+          represented.name.present? && (
+              (!represented.is_public && current_user.allowed_to?(:save_queries, represented.project, global: represented.project.nil?)) ||
+              (represented.is_public && current_user.allowed_to?(:manage_public_queries, represented.project, global: represented.project.nil?))
+            ) && @errors.empty?
+        end
+      end
+    end
```
Comment (ulferts) [inline lib/api/v3/queries/update_form_representer.rb]: Almost the same as CreateFormRepresenter. If possible, please extract the shared code.

Additionally, we still have the duplicate validation checks in both of the representers.

**Review by ulferts — COMMENTED** [review]
Looks good functionality-wise. Please see my separate comments.

**Comment by ulferts** [comment]
Will wait for travis but looks good so far.

**Comment by ulferts** [comment]
The test failures do not occur on my machine.

**Pull request merged**